doc/upgrading.md

```diff
@@ -0,0 +1,40 @@
+# Managing Upgrades/Downgrades
+This guide describes how to upgrade your SPIRE deployment, as well as the compatibility guarantees that SPIRE users can expect.
+
+## SPIRE Versioning
+SPIRE versions are expressed as **x.y.z**, where **x** is the major version, **y** is the minor version, and **z** is the patch version, following Semantic Versioning terminology. The SPIRE project is currently pre-1.0, so minor SPIRE releases act like major releases and patch releases act like minor releases. Despite the fact that the project is pre-1.0, version compatibility guarantees are already in place.
```

Comment:
This reads much better, but raises a question. I chose here to say "act like" in order to communicate the functional differences between one release type and another without _actually_ renaming minor -> major in the context of this document. Thus, the rest of the sections use the term "minor" to mean 0.x.0.

Here, I think you are suggesting that we use the term "major" instead of "minor" throughout this document even though the true major version number is 0.

There is definitely a lot of confusion about what is minor and what is major in the context of SPIRE and how it relates to semver. Here, I tried to take the absolute definitions and apply them, effectively meaning that there are _no_ major releases... but perhaps that's not the right way to go.

I am happy to take your suggestion as-is, but I also want to be explicit about this decision and apply it everywhere going forward. One last consideration is that pre-1.0 behavior _is_ defined by semver, and I believe our current versioning semantics are in line with what semver defines there...

Reply:
I'm very much in favor of using the absolute definitions. I think doing otherwise will be confusing to anybody with experience in versioning schemes. Maybe we say something akin to "SPIRE is pre-1.0, so our compat guarantees are with +/- 1 minor version. After 1.0, our compatability guarantees will align to major versions" and then keep the rest as-is. When we reach 1.0, we can come update the document accordingly.

Reply:
> I'm very much in favor of using the absolute definitions. I think doing otherwise will be confusing to anybody with experience in versioning schemes.

Yes... the more I read through the document with this in mind, the more it feels like defining the "major" position as 0.x.0 will lead to even more confusion.

Taking @sanderson042's comment though (and @azdagron's previous comment), I think it's also clear that the current wording is confusing. Would it make more sense to leave the "minors are like majors" analogy out altogether?

Reply:
Yes, i think it would. I think it would still be worth mentioning that the guarantees layed out in the document are for pre-1.0 only and that the document will be updated when 1.0 lands or something.

Reply:
Ah, I see where I was confused. This has been a productive discussion. I'm on vacation Thursday & Friday, so this'll be my last contribution. This is a complete straw man proposal, but what if we used "big" and "little" instead of major and minor and pared this paragraph down to one sentence. The method to my madness is that big and little don't have preexisting formal definitions like in semver. Anyway, I think you guys have found the confusing bits, so can figure out the wording from here.

"In its current pre-1.0 state, versions in the SPIRE project are expressed as 0.big.little, where big indicates a release version with many changes and little indicates a release with a few changes."

See more below...

Reply:
I chatted around about this one, and the conclusion I reached is that 1) it is the "x is like y" and "y is like z" talk that is the primary source of confusion, and 2) that detail is not very important because the impact it has on upgrade path is explicitly spelled out through the rest of this document.

I spoke to @sanderson042 offline about this, and we decided to simply strike that statement.

I did, however, add one more sentence here stating that we would update this document prior to a 1.0.0 release in an effort to quell any questions that may pop up around how this will work in the future